streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java

```diff
@@ -1378,4 +1374,168 @@ public synchronized void updateSubscriptions(final SubscriptionUpdates subscript
+
+        return description;
+    }
+
+    private void describeSubtopologies(final TopologyDescription description) {
+        for (final Map.Entry<Integer, Set<String>> nodeGroup : makeNodeGroups().entrySet()) {
+
+            final Set<String> allNodesOfGroups = nodeGroup.getValue();
+            final boolean isNodeGroupOfGlobalStores = nodeGroupContainsGlobalSourceNode(allNodesOfGroups);
+
+            if (!isNodeGroupOfGlobalStores) {
+                describeSubtopology(description, nodeGroup.getKey(), allNodesOfGroups);
+            }
+        }
+    }
+
+    private boolean nodeGroupContainsGlobalSourceNode(final Set<String> allNodesOfGroups) {
+        for (final String node : allNodesOfGroups) {
+            if (isGlobalSource(node)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void describeSubtopology(final TopologyDescription description,
+                                     final Integer subtopologyId,
+                                     final Set<String> nodeNames) {
+
+        final HashMap<String, TopologyDescription.Node> nodesByName
+            = addAllNodesWithoutPredecessorsOrSuccessors(nodeNames);
+
+        connectAllNodesToTheirPredecessorsAndSuccessors(nodesByName);
+
+        description.subtopologies.add(new TopologyDescription.Subtopology(
+            subtopologyId,
+            new HashSet<>(nodesByName.values())));
+    }
+
+    private HashMap<String, TopologyDescription.Node> addAllNodesWithoutPredecessorsOrSuccessors(final Set<String> nodeNames) {
+        final HashMap<String, TopologyDescription.Node> nodesByName = new HashMap<>();
+
+        for (final String nodeName : nodeNames) {
+            final NodeFactory nodeFactory = nodeFactories.get(nodeName);
+            final TopologyDescription.Node node;
+
+            if (nodeFactory instanceof SourceNodeFactory) {
+                node = new TopologyDescription.Source(
+                    nodeName,
+                    getSourceTopicListOrPatternAsString(nodeName));
+
+            } else if (nodeFactory instanceof ProcessorNodeFactory) {
+                node = new TopologyDescription.Processor(
+                    nodeName,
+                    getStoresOfNode(nodeName));
+
+            } else if (nodeFactory instanceof SinkNodeFactory) {
+                node = new TopologyDescription.Sink(
+                    nodeName,
+                    nodeToSinkTopic.get(nodeName));
+
+            } else {
+                throw new TopologyBuilderException("Unknown node type: " + nodeFactory.getClass().getName());
+            }
+
+            nodesByName.put(nodeName, node);
+        }
+
+        return nodesByName;
+    }
+
+    private String getSourceTopicListOrPatternAsString(final String sourceName) {
+        final Pattern sourcePattern = nodeToSourcePatterns.get(sourceName);
+
+        if (sourcePattern != null) {
+            return sourcePattern.pattern();
+        } else {
```

Comment:
I personally think, it reads better (at least my mind has an easier time parsing this).